### PR TITLE
API/COMPAT: add pydatetime-style positional args to Timestamp constructor

### LINKED ISSUE #10758: API/COMPAT: add kw args to Timestamp constructor
from [SO](http://stackoverflow.com/questions/21599633/how-to-create-a-pandas-timestamp-object)

``Timestamp(year, month, day, second, millisecond, microsecond, nanosecond)``

as optional keyword args might be nice for compat with ``datetime.datetime``. We do something similar for ``Timedelta`` construction.

### PR DESCRIPTION
**Pull request by thejohnfreeman**
- [X] closes #10758
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

If more tests are desired, please let me know where to put them. Same for the whatsnew entry. This is my first pull request for Pandas.

cc @jreback

### CODE REVIEW

File: pandas/tslib.pyx [hunk 1]
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
...
-    def __new__(cls, object ts_input, object offset=None, tz=None, unit=None):
+    def __new__(cls,
+            # Keep legacy parameter names.
+            object ts_input, object offset=None, tz=None, unit=None,
+            # Switch to datetime positional paramers, starting with `minute`.
+            minute=0, second=0, microsecond=0, tzinfo=None):
         cdef _TSObject ts
         cdef _Timestamp ts_base
 
+        if offset is not None and type(offset) is int:
```
Comment (jreback) [inline pandas/tslib.pyx]: use ``is_integer_object(offset)``

File: pandas/tslib.pyx [hunk 2]
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
     def combine(cls, date, time):
         return cls(datetime.combine(date, time))
 
-    def __new__(cls, object ts_input, object offset=None, tz=None, unit=None):
+    def __new__(cls,
+            # Keep legacy parameter names.
+            object ts_input, object offset=None, tz=None, unit=None,
```
Comment (jreback) [inline pandas/tslib.pyx]: put the comments below the actual constructor

File: pandas/tseries/tests/test_tslib.py [hunk 3]
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
         with tm.assertRaisesRegexp(ValueError, 'Cannot convert Period'):
             Timestamp(Period('1000-01-01'))
 
+    def test_constructor_positional(self):
+        with tm.assertRaises(TypeError):
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: add the issue number here

File: pandas/tseries/tests/test_tslib.py [hunk 4]
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
...
+            Timestamp(2000, 1)
+        with tm.assertRaises(ValueError):
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: not sure if its worth it to try to make ``Timestamp(2000, 1)`` raise a ``ValueError``

File: pandas/tseries/tests/test_tslib.py [hunk 5]
```diff
@@ -177,6 +177,29 @@ def test_constructor_invalid(self):
...
+            Timestamp(2000, 0, 1)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 13, 1)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 0)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 32)
+
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: add some tests where ``None`` is passed for various places ``Timestamp(2000, 1, 1, None, None)``
``Timestamp(2000, 1, None, ...)``

File: pandas/tslib.pyx [hunk 6]
```diff
@@ -291,10 +291,25 @@ class Timestamp(_Timestamp):
...
+            # Switch to datetime positional paramers, starting with `minute`.
+            minute=0, second=0, microsecond=0, tzinfo=None):
```
Comment (jreback) [inline pandas/tslib.pyx]: These should default to ``None`` rather than ``0``, I *think* that ``datetime.datetime`` will handle that

**Comment by jreback** [comment]
pls add a whatsnew entry.

any luck with accepting the kwargs as well? (or too complicated)

File: pandas/tslib.pyx (hunk 6 above)
Comment (thejohnfreeman) [inline pandas/tslib.pyx]: `datetime.datetime` raises `TypeError` when `None` is passed for any argument (except `tzinfo`), which led me to default to `0`. Is that ok?

```
----> 1 dt.datetime(2012, 1, 1, None)

TypeError: an integer is required (got type NoneType)
```

**Comment by thejohnfreeman** [comment]
Should the whatsnew go in for 18.1 or something else?

kwargs got too complicated. I wanted to reach a working version of positional parameters first before building kwargs on top of it. I will try out an idea this weekend.

File: pandas/tslib.pyx (hunk 6 above)
Comment (jreback) [inline pandas/tslib.pyx]: ok.

I would ideally I think like to have this signature:

``ts_input=None, offset=None, tz=None, unit=None, year=None, month=None, day=None, hour=None, minute=None, second=None, tzinfo=None``

then its simple to do kwargs, e.g. you MUST provide ``year`` is not None (and all of ts_input, offset, tz, unit must be ``None``).

Then if you are actually creating a ``datetime.time`` (thru either positional or kwargs), then you can just call like (after you have done a rename of args for the positional if needed).

e.g.

```datetime.datetime(year=year, month=month, day=day, hour=hour or 0, minute=minute or 0, second=second or 0)``

**Comment by jreback** [comment]
you can put in 0.18.1; if you finish up before we release we can just move

File: doc/source/whatsnew/v0.18.1.txt [hunk 7]
```diff
@@ -22,6 +22,14 @@ New features
 Enhancements
 ~~~~~~~~~~~~
 
+- Timestamps accept positional parameters like `datetime.datetime`:
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: put ``Timestamp`` in double back-ticks 

say: can now accept

put a link in for ``datetime.datetime`` back to the constructor (on the python documentation page)

@jorisvandenbossche do we have a nicer way of doing this (as opposed to a straight link), IIRC, sphinx allows us to use an intra-module link (like we do with numpy)

Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.18.1.txt]: yes, you can just do ```:func:`datetime.datetime` ``` and that will link to the standard library's method

**Comment by jreback** [comment]
@thejohnfreeman can you rebase / update & move whatsnew to 0.18.2

**Comment by thejohnfreeman** [comment]
Ok, I rebased. I have time this week to finish this patch. I wrote a longer comment asking a question, but I'll just go ahead and try the signature you gave above.

```
def __new__(cls,
        object ts_input, object offset=None, tz=None, unit=None,
        year=None, month=None, day=None,
        hour=None, minute=None, second=None, tzinfo=None):
```

File: pandas/tslib.pyx [hunk 8]
```diff
@@ -288,10 +289,36 @@ class Timestamp(_Timestamp):
...
         return cls(datetime.combine(date, time))
 
-    def __new__(cls, object ts_input, object offset=None, tz=None, unit=None):
+    def __new__(cls,
+            object ts_input, object offset=None, tz=None, unit=_no_unit,
+            minute=0, second=0, microsecond=0, tzinfo=None):
+        # The parameter list folds together legacy parameter names (the first
+        # four) and positional parameter names from pydatetime (starting with
+        # `minute`).
+        #
```
Comment (jreback) [inline pandas/tslib.pyx]: good, this is a nice way to do this.

File: pandas/tseries/tests/test_tslib.py [hunk 9]
```diff
@@ -180,6 +180,36 @@ def test_constructor_invalid(self):
...
+            Timestamp(2000, 1, 1, None, None, None)
+
+        ts = Timestamp(2000, 1, 2)
+
+        actual = ts.to_pydatetime()
+        expected = datetime.datetime(2000, 1, 2)
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
+
+        pos_args = [2000, 1, 2, 3, 4, 5, 999999]
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: can u also add another with ``tzinfo`` (another set of pos_args).

and another with kwargs as well.

**Comment by jorisvandenbossche** [comment]
@thejohnfreeman You will also need to update the Timestamp docstring

**Comment by jreback** [comment]
also I believe its trivial to close #11630 (or it might already be done). Add a test and should be straightforward.

**Comment by thejohnfreeman** [comment]
Take a look at my attempt at supporting both positional and keyword arguments. I removed some tests for TypeError; it just means we are more permissive. I'll fix up the doc string and add the other requested tests tomorrow.

File: pandas/tslib.pyx [hunk 10]
```diff
@@ -288,10 +288,43 @@ class Timestamp(_Timestamp):
...
+
         cdef _TSObject ts
         cdef _Timestamp ts_base
 
+        if ts_input is None:
+            # User passed keyword arguments.
+            return Timestamp(datetime(year, month, day, hour or 0,
+                minute or 0, second or 0, microsecond or 0, tzinfo),
+                tz=tzinfo)
+        if is_integer_object(offset):
```
Comment (jreback) [inline pandas/tslib.pyx]: elif

**Comment by jreback** [comment]
whoosh @thejohnfreeman nice constuctor now! yeah prob need some more tests.

**Comment by shoyer** [comment]
Effectively, we are doing type dependent multiple dispatch here with the `Timestamp` constructor. To be honest, I'm not a big fan of adding multiple paths like this that do such different things with the arguments.

**Comment by jreback** [comment]
@shoyer that doesn't make any sense. ``Timestamp`` is a sub-class of ``datetime``. It doesn't take the correct (positional, nor keyword args ATM). I don't see the problem here. Yes it also does a lot more, but that's exactly the point, it IS user friendly w/o be ambiguous.

**Comment by thejohnfreeman** [comment]
I updated the docstring and added tests. I'm considering this a first draft submitted for approval. If it's approved, I'll squash the commits and we can wait for Travis.

File: pandas/tseries/tests/test_tslib.py [hunk 11]
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
...
+            Timestamp(2000, 1, 0)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 32)
+
+        ts = Timestamp(2000, 1, 2)
+
+        actual = ts.to_pydatetime()
+        expected = datetime.datetime(2000, 1, 2)
+        self.assertEqual(actual, expected)
+        self.assertEqual(type(actual), type(expected))
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: what is the purpose of this test here? ``.to_pydatetime()`` is already tested elsewhere.

File: pandas/tslib.pyx [hunk 12]
```diff
@@ -225,8 +225,12 @@ class Timestamp(_Timestamp):
     for the entries that make up a DatetimeIndex, and other timeseries
     oriented data structures in pandas.
 
-    Parameters
-    ----------
+    There are essentially three calling conventions for the constructor. The
+    first, legacy form accepts four parameters. They can be passed by
+    position or keyword.
+
+    Legacy Parameters
```
Comment (jreback) [inline pandas/tslib.pyx]: don't call these ``Legacy`` Parameters; these are in fact the normal calling convention. The ``datetime.datetime`` convention are the compat parameters (but you don't have to call it differently).

File: pandas/tslib.pyx [hunk 13]
```diff
@@ -235,6 +239,20 @@ class Timestamp(_Timestamp):
         Time zone for time which Timestamp will have.
     unit : string
         numpy unit used for conversion, if ts_input is int or float
+
+    The other two forms copy the parameters from datetime.datetime. They can
+    be passed by either position or keyword, but not both mixed together.
+
+    datetime.datetime Parameters
+    ------------------------------
```
Comment (jreback) [inline pandas/tslib.pyx]: put a refernce to ``datetime.datetime`` doc page

File: pandas/tseries/tests/test_tslib.py [hunk 14]
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
         with tm.assertRaisesRegexp(ValueError, 'Cannot convert Period'):
             Timestamp(Period('1000-01-01'))
 
+    def test_constructor_positional(self):
+        # GH 10758
+        with tm.assertRaises(TypeError):
+            Timestamp(2000, 1)
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: can you use ``assertRaisesRegexp`` here to check for things

```
 In [1]: datetime.datetime(2000,1)
TypeError: Required argument 'day' (pos 3) not found
```

doesn't have to be an exact checking, but need to make sure that it is raising the ``datetime.datetime`` constructor error message and not something else.

File: pandas/tseries/tests/test_tslib.py [hunk 15]
```diff
@@ -180,6 +180,63 @@ def test_constructor_invalid(self):
...
+            Timestamp(year=2000, month=13, day=1)
+        with tm.assertRaises(ValueError):
+            Timestamp(year=2000, month=1, day=0)
+        with tm.assertRaises(ValueError):
+            Timestamp(year=2000, month=1, day=32)
+
+        ts = Timestamp(year=2000, month=1, day=2)
+
+        actual = ts.to_pydatetime()
+        expected = datetime.datetime(year=2000, month=1, day=2)
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: this also seems superfluous (the ``.to_pydatetime()`` check)

**Comment by jreback** [comment]
did you add a test for #11630 as well (I *think* you covered it in code, just need the test)

**Comment by thejohnfreeman** [comment]
For #11630, I'm considering the test in test_constructor_positional to be equivalent:

```
ts = Timestamp(2000, 1, 2)
```

I can add a test with the exact parameters from #11630 too. Regardless, I will add the issue in a comment.

File: pandas/tseries/tests/test_tslib.py (hunk 11 above)
Comment (thejohnfreeman) [inline pandas/tseries/tests/test_tslib.py]: To test for equivalence. Let me know if I should use a different pattern.

File: pandas/tslib.pyx (hunk 13 above)
Comment (thejohnfreeman) [inline pandas/tslib.pyx]: What's the format for that?

File: pandas/tseries/tests/test_tslib.py (hunk 11 above)
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: I understand but that's not part of this PR (and its already tested), so not necessary.

**Comment by jreback** [comment]
@thejohnfreeman yes add that exact test (e.g. 

```
self.assertEqual(repr(Timestamp(2015, 11, 12)),  repr(Timestamp('20151112')))
```
should be enough

File: doc/source/whatsnew/v0.18.2.txt [hunk 16]
```diff
@@ -31,6 +31,14 @@ Other enhancements
 
 - The ``.tz_localize()`` method of ``DatetimeIndex`` and ``Timestamp`` has gained the ``errors`` keyword, so you can potentially coerce nonexistent timestamps to ``NaT``. The default behaviour remains to raising a ``NonExistentTimeError`` (:issue:`13057`)
 
+- ``Timestamp``s can now accept positional parameters like :func:`datetime.datetime`:
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: add the issue numbers at the end here

File: pandas/tslib.pyx [hunk 17]
```diff
@@ -225,8 +225,12 @@ class Timestamp(_Timestamp):
     for the entries that make up a DatetimeIndex, and other timeseries
     oriented data structures in pandas.
 
-    Parameters
-    ----------
+    There are essentially three calling conventions for the constructor. The
+    first, legacy form accepts four parameters. They can be passed by
+    position or keyword.
```
Comment (jreback) [inline pandas/tslib.pyx]: add a versionadded tag here (0.18.2)

File: pandas/tslib.pyx (hunk 13 above)
Comment (jreback) [inline pandas/tslib.pyx]:
```
    .. versionadded:: 0.18.2
```

be sure to put a blank line before and after (otherwise we get warnings)

**Comment by shoyer** [comment]
> @shoyer that doesn't make any sense. Timestamp is a sub-class of datetime. It doesn't take the correct (positional, nor keyword args ATM). I don't see the problem here. Yes it also does a lot more, but that's exactly the point, it IS user friendly w/o be ambiguous.

You cannot add alternate signatures without adding ambiguity. For example, suppose a naive user tries to pass three string arguments instead of integers into the new constructor. This is invalid for both datetime and the existing Timestamp signature, but because these arguments are supplied as strings, it would still trigger the existing codepath.

Let's compare what they would see:
```
>>> pd.Timestamp('2000', '01', '01')
UnknownTimeZoneError: '01'

>>> datetime.datetime('2000', '01', '01')
TypeError: an integer is required
```
The error from datetime is *much* more informative.

The standard library `datetime.datetime` class does not accept a string to convert to avoid exactly these sorts of issues. Instead, you are forced to use alternate constructors such as `strptime`.

It is unambiguously poor API design to make functions that do two very different things depending on the exact form of the input.

I'm all for making `pd.Timestamp` more consistent with `datetime.datetime` if you're willing to deal with the hassle of a long deprecation cycle, but the right approach is not to overload the constructor to do something totally different. The right approach would be to switch the constructor entirely.

**Comment by thejohnfreeman** [comment]
Personally, I can sympathize with @shoyer. In all languages, I prefer that constructors just directly initialize fields and leave any magic to ordinary functions. Especially in Python, you only get one constructor but you can make unlimited functions for every style of construction you want, and it becomes easier to create custom test objects.

That said, I think it's more important to fit into a project's existing style and API, and I can see an argument for fitting this functionality into the `Timestamp` constructor. For now, I'm going to follow Jeff's instructions on this since that's what I've been doing all along.

File: pandas/tseries/tests/test_tslib.py (hunk 14 above)
Comment (thejohnfreeman) [inline pandas/tseries/tests/test_tslib.py]: For the cases where not enough parameters are passed to `Timestamp`, the error will be `an integer is required (got NoneType)` instead of `Required argument not found`. Matching the `datetime` exception exactly requires 3 additional checks and calls:

```
if year is None:
    return datetime.datetime()
if month is None:
    return datetime.datetime(year)
if day is None:
    return datetime.datetime(year, month)
```

which seems ugly. Do we really want that?

**Comment by shoyer** [comment]
What about deprecating passing a single object (string or datetime) to `pd.Timestamp` in favor of using `pd.to_datetime`? That way we could get back to only one constructor eventually.

It would not be the end of the world to add this alternate constructor but IMO it would not be a positive improvement to have both.

If to do do it, let's clearly document the two constructors at the top of the docstring for the class, e.g.,
```
Timestamp(ts_input, object_offset=None, tz=None, unit=None)
Timestamp(year, month, day[, hour[, minute[, second[, microsecond[, tzinfo]]]]])

Timestamp is...
```
I *think* sphinx can process that correctly but we should double check.

**Comment by jreback** [comment]
@shoyer why would you blow up a very useful interface?

``Timestamp('20130101')`` is such a useful feature and you want to DEPRECATE?

the entire point is to REDUCE mental effort here. The whole notion of a user having to choose alternate constructors (and I know we had this discussion w.r.t. ``RangeIndex``) is completely bonkers to me.

Sure if things are ambiguous, then you have to choose. In this case it is wholly unambiguous, replicates the existing (python ``datetime.datetime`` API), and provides very useful parsing abilities, not to mention its very light weight.

**Comment by shoyer** [comment]
> Timestamp('20130101') is such a useful feature and you want to DEPRECATE?

I agree -- my proposal is not entirely serious. I do think we need to make hard choices like this to achieve reliable APIs.

>the entire point is to REDUCE mental effort here. The whole notion of a user having to choose alternate constructors (and I know we had this discussion w.r.t. RangeIndex) is completely bonkers to me.

Just to prove I'm not crazy, here are some high lights from the first page of a Google search for "Python multiple constructors":
http://stackoverflow.com/questions/2164258/multiple-constructors-in-python
http://stackoverflow.com/questions/682504/what-is-a-clean-pythonic-way-to-have-multiple-constructors-in-python
http://www.erol.si/2016/01/python-and-multiple-constructors/
https://pythonconquerstheuniverse.wordpress.com/2010/03/17/multiple-constructors-in-a-python-class/ (read the comments)

I could go on, but you get the point.

You'll notice that the universal advice seems to be to skip the magic and use classmethods for alternate constructors instead of overloading `__init__`.

**Comment by kawochen** [comment]
If we are to add more code paths in the constructor, is it possible to make the entry points more obvious by calling specific `from_*` methods inside `__new__` or `__init__`?  After reading the doc string I feel like it might become harder for a casual contributor like me to debug the code.  But I think this can be remedied by making the intention of each code path more visible in the code.

**Comment by thejohnfreeman** [comment]
Hi @kawochen, I too want to make the code paths understandable. I had added some comments to each new path to document its intention. Did they help clarify at all?

**Comment by thejohnfreeman** [comment]
I keep getting lint errors in the automated check. It doesn't say what the error is, and running trace8 locally doesn't give me any indication either.

```
The command "ci/script.sh" exited with 1.
1.00s$ ci/lint.sh
inside ci/lint.sh
discarding /Users/travis/miniconda/envs/pandas/bin from PATH
prepending /Users/travis/miniconda/envs/pandas/bin to PATH
NOT Linting
```

Any ideas what I need to do to get Travis to pass?

**Comment by jreback** [comment]
it's not lint errors but a failed test

https://travis-ci.org/pydata/pandas/jobs/129782361

**Comment by thejohnfreeman** [comment]
Ah, it expects `Timestamp(None)` to return Not a Time (`NaT`).

File: pandas/tslib.pyx [hunk 18]
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
...
+
+    :func:`datetime.datetime` Parameters
+    ------------------------------
+
+    .. versionadded:: 0.18.2
+
+    year : int
+    month : int
+    day : int
+    hour : int [optional]
```
Comment (jreback) [inline pandas/tslib.pyx]: I think these need to be: ``int, optional`` (of this form)

File: pandas/tslib.pyx [hunk 19]
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
         Time zone for time which Timestamp will have.
     unit : string
         numpy unit used for conversion, if ts_input is int or float
+
+    The other two forms mimic the parameters from datetime.datetime. They can
+    be passed by either position or keyword, but not both mixed together.
+
+    :func:`datetime.datetime` Parameters
+    ------------------------------
```
Comment (jreback) [inline pandas/tslib.pyx]: make this line extend the entire length (to end of text)

File: pandas/tslib.pyx [hunk 20]
```diff
@@ -235,6 +240,23 @@ class Timestamp(_Timestamp):
         Time zone for time which Timestamp will have.
     unit : string
         numpy unit used for conversion, if ts_input is int or float
+
+    The other two forms mimic the parameters from datetime.datetime. They can
```
Comment (jreback) [inline pandas/tslib.pyx]: put ``datetime.datetime`` in double-backticks

File: pandas/tslib.pyx [hunk 21]
```diff
@@ -288,10 +310,43 @@ class Timestamp(_Timestamp):
...
+            object ts_input=_no_input, object offset=None, tz=None, unit=None,
+            year=None, month=None, day=None,
+            hour=None, minute=None, second=None, microsecond=None,
+            tzinfo=None):
+        # The parameter list folds together legacy parameter names (the first
+        # four) and positional and keyword parameter names from pydatetime.
+        #
+        # There are three calling forms:
+        # - In the legacy form, the first parameter, ts_input, is required
+        # and may be datetime-like, str, int, or float. The second parameter,
```
Comment (jreback) [inline pandas/tslib.pyx]: can you format this like:

```
# - In the ......
#   next line here
#   next line here
....
```

File: pandas/tseries/tests/test_tslib.py [hunk 22]
```diff
@@ -180,6 +180,50 @@ def test_constructor_invalid(self):
...
+
+        # GH 11630
+        self.assertEqual(
+            repr(Timestamp(2015, 11, 12)),
+            repr(Timestamp('20151112')))
+
+        self.assertEqual(
+            repr(Timestamp(2015, 11, 12, 1, 2, 3, 999999)),
+            repr(Timestamp('2015-11-12 01:02:03.999999')))
+
```
Comment (jreback) [inline pandas/tseries/tests/test_tslib.py]: can you add the assertion for ``Timestamp(None)`` as  well

**Comment by jreback** [comment]
lgtm. @thejohnfreeman nice explanations in the comments. I think it is clear.

@jorisvandenbossche @shoyer 

comments?

**Comment by jreback** [comment]
@thejohnfreeman did you add the test for #11630 (e.g the ``repr`` expr from above)

File: pandas/tseries/tests/test_tslib.py [hunk 23]
```diff
@@ -180,6 +180,52 @@ def test_constructor_invalid(self):
...
+            Timestamp(2000, 13, 1)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 0)
+        with tm.assertRaises(ValueError):
+            Timestamp(2000, 1, 32)
+
+        # GH 11630
+        self.assertEqual(
+            repr(Timestamp(2015, 11, 12)),
+            repr(Timestamp('20151112')))
```
Comment (thejohnfreeman) [inline pandas/tseries/tests/test_tslib.py]: Here is the test for #11630, Github didn't do the best job linking it to your earlier comment.

**Comment by jreback** [comment]
thanks @thejohnfreeman great PR!

**Comment by thejohnfreeman** [comment]
Thank you!
